causalml/inference/tree/bart.py

```diff
@@ -0,0 +1,547 @@
+                self.trees[j] = new_tree
+
+            y_pred = self.predict(X, train=True) # opportunity to parallelize according to Pratola et al. (2013) (https://arxiv.org/abs/1309.1906)
+            self.sum_of_trees = y_pred
+            if self.task == 'regression':
+                squared_error = (transformed_y.reshape(-1, 1) - y_pred)**2
+                p_alpha = (alpha_ig + len(y_pred)) / 2
+                p_beta = (beta_ig + squared_error.sum()) / 2
+                sigma = 1. / np.sqrt(np.random.gamma(p_alpha, 1./p_beta))
+
+    def _predict(self, X:np.ndarray, rescale:bool) -> np.ndarray:
+        """Private method for prediction that works equally for Classic ML and Uplift Modeling
+        """
+        # predict for each tree and aggregate
+        predictions = np.zeros((X.shape[0], 1))
+        for tree in self.trees:
+            predictions += tree.predict(X)
+            
+        if self.task == 'classification':
+            return self.sigmoid(predictions)
+        
+        if rescale:
+            predictions = self.rescale_y(predictions)
+        
+        return predictions
+    
+    def predict(self, X:np.ndarray, train:bool = False, rescale:bool = False, full_output:bool = False) -> np.ndarray:
+        """Predict treatment effects if the problem type is uplift modeling or one-dimensional response if 
+        problem type is classic ml.
+        Args:
+            X (np.array): A feature matrix
+            train (bool): Optional. To be used only by private methods. Distinguishes if prediction to be done for
+                            each treatment in case of uplift modeling setting.
+            rescale (bool): True if output wanted to be in the same scale as the original y.
+            full_output (bool): True if the predictions for each treatment, alongside with the deltas with control
+                                    as baseline. False if only deltas to be returned.
+        Returns:
+            (pd.DataFrame): Predictions of treatment effects if the problem type is uplift modeling or one-dimensional 
+                                response if problem type is classic ml.
+        """
+        if self.problem_type == 'uplift_modeling' and not train:
+            columns = []
+            predictions = []
+            for treatment in self.treatments:
+                treatment_predictions = self._predict(np.concatenate([X, np.ones((len(X), 1)) * treatment], axis=1), rescale)
+                predictions.append(treatment_predictions.flatten().tolist())
+                columns.append(f'treatment_{treatment}' if treatment > 0 else 'control')
+            
+            df_res = pd.DataFrame(np.array(predictions).T, columns=columns)
+            
+            # From: https://github.com/uber/causalml/blob/c42e873061eb74ec9c3ca6ea991e113b886245ae/causalml/inference/tree/uplift.pyx
+            df_res['recommended_treatment'] = df_res.apply(np.argmax, axis=1)
```

Comment:
Not necessarily something to solve here, but we need to change the terminology of "recommended treatment", because CATE is just one factor that is relevant for whether a treatment should be recommended or not.